### PR TITLE
Fix remote command execution behavior

### PR DESCRIPTION
**Pull request by holesch**
Makes executing remote commands less surprising.

### CODE REVIEW

File: fallbear-cmd [hunk 1]
```diff
@@ -2,63 +2,85 @@
...
+echo_debug "PVUSER: $PVUSER"
+
+# get home directory and login shell from passwd entry, format is:
+# name:password:UID:GID:GECOS:directory:shell
+IFS=":" read -r _ _ _ _ _ home shell 2>/dev/null << EOF
+$(nsenter --target "$p" --mount -- grep "^$PVUSER:" /etc/passwd)
+EOF
+
+echo_debug "shell: $shell"
+echo_debug "home: $home"
```
Comment (asac) [inline fallbear-cmd:57]: maybe move these one below so we see the actual final values used if debug is on.

File: fallbear-cmd [hunk 2]
```diff
@@ -2,63 +2,85 @@
...
+
+# set default values if no passwd entry was found
+: "${shell:=/bin/sh}"
+: "${home:=/tmp1}"
```
Comment (asac) [inline fallbear-cmd]: to make this less "magic" for the opportunistic contributor maybe use a pattern like home=${home:-/tmp1}

File: fallbear-cmd [hunk 3]
```diff
@@ -2,63 +2,85 @@
...
-pwdline=`nsenter -m/proc/$p/ns/mnt -- [ -f /etc/passwd ] \
-		 && (nsenter -m/proc/$p/ns/mnt cat /etc/passwd | grep ^$PVUSER:) \
-		 || true`
+echo_debug "PVUSER: $PVUSER"
+
+# get home directory and login shell from passwd entry, format is:
+# name:password:UID:GID:GECOS:directory:shell
+IFS=":" read -r _ _ _ _ _ home shell 2>/dev/null << EOF
+$(nsenter --target "$p" --mount -- grep "^$PVUSER:" /etc/passwd)
+EOF
```
Comment (asac) [inline fallbear-cmd:50]: nice pattern :)

File: fallbear-cmd [hunk 4]
```diff
@@ -2,63 +2,85 @@
...
+# Use shell quoting magic to preserve arguments with whitespace, while changing
+# to the home directory first: Arguments are only passed correctly with $@ in
+# double quotes. To pass arguments to sh, the following form is used:
+#
+#    sh -c command_string [argv0 [argument...]]
+#
+# where argv0 is set to its default value "sh". The arguments to exec are in
+# single quotes, so that the called sh process expands those arguments instead
+# of the calling (i.e. this) process.
+set -- sh -c "[ -x '$home' ] && cd '$home'; "'exec "$@"' sh "$@"
```
Comment (asac) [inline fallbear-cmd:82]: took me a moment to parse this ... but I can't think of an easier way to do the same exec magic.... so unless you know a different way lets keep it as you propose.

File: fallbear-cmd [hunk 5]
```diff
@@ -2,63 +2,85 @@
 
 echo_debug() {
     if [ -n "$PVDEBUG" ]; then
-	echo $@
+        echo "$@" >&2
```
Comment (asac) [inline fallbear-cmd]: we usually use tabs not whitespaces to indent in shell and C

File: fallbear-cmd [hunk 6]
```diff
@@ -2,63 +2,85 @@
...
-
 if [ "$plat" = "$plat_tty" ]; then
-	plat_tty=
+    plat_tty=
 fi
 
-if ! [ "$plat" = "/" -o "$plat" = "_pv_"  ]; then
-    p=`lxc_pid $plat`
+case "$plat" in
+"/"|"_pv_")
```
Comment (asac) [inline fallbear-cmd]: not an issue, but elsewhere in our code base we indent the case matches once and the code + ;; twice

**Review by asac — APPROVED** [review]

File: fallbear-cmd (hunk 4 above)
Comment (holesch) [reply fallbear-cmd:82]: I know, it's hard to read and I don't know how to do it differently. That's why I added the comment, to try to explain what's going on.

File: fallbear-cmd (hunk 5 above)
Comment (holesch) [reply fallbear-cmd]: Will fix it. The original was mixed, so I didn't know.

**Pull request merged**